streams/src/main/java/org/apache/kafka/streams/kstream/internals/KStreamImpl.java

```diff
@@ -1271,4 +1279,71 @@ public void process(final ProcessorSupplier<? super K, ? super V> processorSuppl
 
         builder.addGraphNode(streamsGraphNode, processNode);
     }
+
+    @Override
+    public KTable<K, V> toTable() {
+        return toTable(NamedInternal.empty());
+    }
+
+    @Override
+    public KTable<K, V> toTable(Named named) {
+        Objects.requireNonNull(named, "named can't be null");
+        final MaterializedInternal<K, V, KeyValueStore<Bytes, byte[]>> materializedInternal =
+            new MaterializedInternal<>(
+                Materialized.with(keySerde, valSerde),
+                builder,
+                TO_KTABLE_NAME);
+
+        return toTable(named, materializedInternal);
+    }
+
+    @Override
+    public <VR> KTable<K, VR> toTable(Materialized<K, VR, KeyValueStore<Bytes, byte[]>> materialized) {
+        return toTable(NamedInternal.empty(), materialized);
+    }
+
+    @Override
+    public <VR> KTable<K, VR> toTable(Named named, Materialized<K, VR, KeyValueStore<Bytes, byte[]>> materialized) {
+        Objects.requireNonNull(named, "named can't be null");
+
+        final MaterializedInternal<K, VR, KeyValueStore<Bytes, byte[]>> materializedInternal =
+            new MaterializedInternal<>(materialized);
+        return toTable(named, materializedInternal);
+    }
+
+    private <VR> KTable<K, VR> toTable(Named named,
+                                       MaterializedInternal<K, VR, KeyValueStore<Bytes, byte[]>> materializedInternal) {
+        final String name = new NamedInternal(named).orElseGenerateWithPrefix(builder, TO_KTABLE_NAME);
+
+        final String storeName = materializedInternal.storeName() == null
+            ? builder.newStoreName(TO_KTABLE_NAME)
+            : materializedInternal.storeName();
```

Comment:
This seems to always enforce a materialization, but I think we should materialize only if we need to.